cmake/functions.cmake

```diff
@@ -0,0 +1,92 @@
+        # The content name. Also used as the package name for find_package().
+        name
+        # Git tag, branch name or hash. If the value has the format
+        # [v]major[.minor[.patch[.tweak]]] and the SKIP_FIND option is not
+        # specified, use find_package() first.
+        # The variable GC_<NAME>_VERSION could be used to override this argument.
+        git_tag_or_version
+        # Git repository URL. If the variable GC_<NAME>_SRC_DIR is defined,
+        # the local directory is used instead.
+        git_repository
+        #
+        # Optional arguments:
+        # SKIP_ADD: Populate but do not add the content to the project.
+        # SKIP_FIND: Do not use find_package().
+        # CMAKE_ARGS: Passed to FetchContent_Declare.
+)
+    string(TOUPPER ${name} uname)
+    cmake_parse_arguments(GC_${uname} "SKIP_ADD;SKIP_FIND" "" "CMAKE_ARGS" ${ARGN})
+
+    if (DEFINED GC_${uname}_CMAKE_ARGS)
+        message(STATUS "${name}_CMAKE_ARGS: ${GC_${uname}_CMAKE_ARGS}")
+    endif ()
+
+    if (DEFINED GC_${uname}_SRC_DIR)
+        FetchContent_Declare(
+                ${name}
+                SOURCE_DIR ${GC_${uname}_SRC_DIR}
+                CMAKE_ARGS ${${uname}_CMAKE_ARGS}
+        )
+    else ()
+        if (DEFINED GC_${uname}_VERSION)
+            set(git_tag_or_version ${GC_${uname}_VERSION})
+        endif ()
+        message(STATUS "${name}_VERSION: ${git_tag_or_version}")
+
+        if (NOT ${uname}_SKIP_FIND AND git_tag_or_version
+                MATCHES "^v?([0-9]+(\.[0-9]+(\.[0-9]+(\.[0-9]+)?)?)?)$")
+            string(REPLACE "v" "" version ${git_tag_or_version})
+            find_package(${name} ${version} EXACT)
+        endif ()
+
+        if (NOT DEFINED ${name}_VERSION_COUNT)
+            message(STATUS "Fetching ${name} from ${git_repository}")
+            FetchContent_Declare(
+                    ${name}
+                    GIT_REPOSITORY ${git_repository}
+                    GIT_TAG ${git_tag_or_version}
+                    CMAKE_ARGS ${GC_${uname}_CMAKE_ARGS}
+                    FIND_PACKAGE_ARGS ${FIND_PACKAGE_ARGS}
+            )
+        endif ()
+    endif ()
+
+    if (NOT DEFINED ${name}_VERSION_COUNT)
+        if (GC_${uname}_SKIP_ADD)
+            if (NOT DEFINED ${name}_POPULATED)
+                FetchContent_Populate(${name})
+                FetchContent_GetProperties(${name})
+                set(${name}_POPULATED TRUE PARENT_SCOPE)
+                set(${name}_SOURCE_DIR ${${name}_SOURCE_DIR} PARENT_SCOPE)
+                set(${name}_BINARY_DIR ${${name}_BINARY_DIR} PARENT_SCOPE)
+            endif ()
+        else ()
+            FetchContent_MakeAvailable(${name})
+        endif ()
+    endif ()
+endfunction()
+
+# Add sources to the global list GC_LIB_SOURCES.
+# The sources could be specified as a list of files or a GLOB pattern:
+#   gc_add_lib_sources(GLOB "src/*.cpp")
+#   gc_add_lib_sources("src/foo.cpp" "src/bar.cpp")
+function(gc_add_lib_sources sources)
```

Comment:
So far we have a huge function that adds two source files to a single library. Do we really need it?

Reply:
For only 2 sources - definitely not. But I think there will be more sources and more subdirectories.  The idea is to use this function in subdirectories to add sources to the lib.